src/main/java/org/kiwiproject/dropwizard/error/dao/jdk/JdbcApplicationErrorDao.java

```diff
@@ -0,0 +1,322 @@
+        var sql = "select * from application_errors" +
+                " where resolved = false and description = ? order by updated_at desc";
+
+        try (var conn = connection(); var ps = conn.prepareStatement(sql)) {
+            ps.setString(1, description);
+            return collectErrors(ps, DEFAULT_PAGE_SIZE);
+        } catch (SQLException e) {
+            throw new ApplicationErrorJdbcException(e);
+        }
+    }
+
+    @Override
+    public List<ApplicationError> getUnresolvedErrorsByDescriptionAndHost(String description, String hostName) {
+        var sql = "select * from application_errors" +
+                " where resolved = false and description = ? and host_name = ? order by updated_at desc";
+
+        try (var conn = connection(); var ps = conn.prepareStatement(sql)) {
+            ps.setString(1, description);
+            ps.setString(2, hostName);
+            return collectErrors(ps, DEFAULT_PAGE_SIZE);
+        } catch (SQLException e) {
+            throw new ApplicationErrorJdbcException(e);
+        }
+    }
+
+    private static List<ApplicationError> collectErrors(PreparedStatement ps, int pageSize) throws SQLException {
+        try (var rs = ps.executeQuery()) {
+            return collectErrors(rs, pageSize);
+        }
+    }
+
+    private static List<ApplicationError> collectErrors(ResultSet rs, int pageSize) throws SQLException {
+        var errors = new ArrayList<ApplicationError>(pageSize);
+        while (rs.next()) {
+            var error = ApplicationErrorJdbc.mapFrom(rs);
+            errors.add(error);
+        }
+        return Collections.unmodifiableList(errors);
+    }
+
+    @Override
+    public long insertError(ApplicationError newError) {
+        checkArgumentIsNull(newError.getId(), "Cannot insert an ApplicationError that has an id");
+
+        var sql = "insert into application_errors" +
+                " (description, exception_type, exception_message, exception_cause_type, exception_cause_message," +
+                " stack_trace, host_name, ip_address, port)" +
+                " values (?, ?, ?, ?, ?, ?, ?, ?, ?)";
+
+        try (var conn = connection(); var ps = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
+            ps.setString(1, newError.getDescription());
+            ps.setString(2, newError.getExceptionType());
+            ps.setString(3, newError.getExceptionMessage());
+            ps.setString(4, newError.getExceptionCauseType());
+            ps.setString(5, newError.getExceptionCauseMessage());
+            ps.setString(6, newError.getStackTrace());
+            ps.setString(7, newError.getHostName());
+            ps.setString(8, newError.getIpAddress());
+            ps.setInt(9, newError.getPort());
+
+            var count = ps.executeUpdate();
+            checkState(count == 1, "Insert count should be one, but is: %s", count);
+
+            try (ResultSet generatedKeys = ps.getGeneratedKeys()) {
```

Comment:
Do we have a utility in kiwi to get generated keys? I know we have a test utility in kiwi-test to do this for JDBI but perhaps we should also have a more general utility. Or, see if Apache Commons has something like this.

Reply:
kiwi-test has the `Jdbi3GeneratedKeys` utility which is specific to JDBI, and is not accessible to production code since kiwi-test is a test library.

Created a [discussion](https://github.com/kiwiproject/kiwi/discussions/1073) in kiwi to consider adding this.